Add prototype string matcher API

achristensen07 opened the pull request:
#### 5e5036cfc5f23e8cc1381e199696a67afa2e62b3
<pre>
Add prototype string matcher API
<a href="https://bugs.webkit.org/show_bug.cgi?id=300829">https://bugs.webkit.org/show_bug.cgi?id=300829</a>
<a href="https://rdar.apple.com/162715982">rdar://162715982</a>

Reviewed by Ryosuke Niwa.

Safari has some sets of strings it searches for as substrings of longer strings.
It needs to do the searches quickly with as little startup time as possible.
Arno experimented with using regular expressions instead, but that
had too much start up time because there are thousands of strings being searched
for.  I experimented with using a typical DFA with one transition per code unit,
but that had 200x as many transitions as this and didn&apos;t fit as well into a CPU&apos;s
L1 cache.  Instead, this PR introduces a trie-based finite state machine that starts
at the root with every code point in the input string.  I slightly changed the format
of the DFA from what is currently in use in Safari to create a primitive that can match
substrings of each other, such as matching &quot;aba&quot; and &quot;ba&quot; and reporting that both
were matched instead of only reporting the shortest one being matched.

This new primitive doesn&apos;t actually enable anything that can&apos;t functionally be done
with regular expressions, but for large sets of regular expressions that are used
over and over again in one process then another, this allows a reduction in time
spent compiling the regular expressions.

Even less useful than something that doesn&apos;t enable new behavior is a native API
that allows string matching on strings.  I added such an interface,
resultsForMatchingCharacters:length:matchAll:searchReverse:, but only temporarily
as a way for Safari to incrementally transition away from the injected bundle use
by having a native matcher that behaves exactly the same as _WKStringMatcher.
Once the transition is complete, this will be removed.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::shouldHaveWebKitNamespaceForWorld):
* Source/WebCore/page/UserContentController.h:
* Source/WebCore/page/UserContentProvider.h:
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::WebKitNamespace):
(WebCore::WebKitNamespace::messageHandlers):
(WebCore::WebKitNamespace::stringMatchers):
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebCore/page/WebKitStringMatcher.cpp: Added.
(WebCore::match):
(WebCore::WebKitStringMatcher::match):
(WebCore::WebKitStringMatcher::Trie::Node::traverse const):
(WebCore::WebKitStringMatcher::Trie::Trie):
(WebCore::WebKitStringMatcher::Trie::serialize const):
(WebCore::WebKitStringMatcher::dataForMatchingStrings):
(WebCore::WebKitStringMatcher::statesAndTransitionsFromVersionedData):
* Source/WebCore/page/WebKitStringMatcher.h: Copied from Source/WebKit/Shared/WebUserContentControllerDataTypes.h.
* Source/WebCore/page/WebKitStringMatcher.idl: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebCore/page/WebKitStringMatcherOptions.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebCore/page/WebKitStringMatcherOptions.idl: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebCore/page/WebKitStringMatchersNamespace.cpp: Copied from Source/WebKit/Shared/WebUserContentControllerDataTypes.h.
(WebCore::WebKitStringMatchersNamespace::WebKitStringMatchersNamespace):
(WebCore::WebKitStringMatchersNamespace::namedItem):
(WebCore::WebKitStringMatchersNamespace::supportedPropertyNames const):
(WebCore::WebKitStringMatchersNamespace::isSupportedPropertyName):
* Source/WebCore/page/WebKitStringMatchersNamespace.h: Copied from Source/WebKit/Shared/WebUserContentControllerDataTypes.h.
(WebCore::WebKitStringMatchersNamespace::create):
* Source/WebCore/page/WebKitStringMatchersNamespace.idl: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/UserContentControllerParameters.h:
* Source/WebKit/Shared/UserContentControllerParameters.serialization.in:
* Source/WebKit/Shared/WebUserContentControllerDataTypes.cpp: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
(WebKit::WebStringMatcherData::WebStringMatcherData):
(WebKit::WebStringMatcherData::sharedMemoryHandle const):
* Source/WebKit/Shared/WebUserContentControllerDataTypes.h:
* Source/WebKit/Shared/WebUserContentControllerDataTypes.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIStringMatcher.cpp: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
(API::StringMatcher::StringMatcher):
(API::StringMatcher::sharedMemory):
(API::StringMatcher::sharedBuffer):
* Source/WebKit/UIProcess/API/APIStringMatcher.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
(-[WKUserContentController _addStringMatcher:contentWorld:name:]):
(-[WKUserContentController _removeStringMatcherWithName:contentWorld:]):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentControllerPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKStringMatcher.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKStringMatcher.mm: Added.
(-[_WKStringMatcher initWithData:]):
(-[_WKStringMatcher initWithDataInFile:]):
(-[_WKStringMatcher dealloc]):
(+[_WKStringMatcher matcherDataForStringsAndIdentifiers:]):
(-[_WKStringMatcher resultsForMatchingCharacters:length:matchAll:searchReverse:]):
(-[_WKStringMatcher _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKStringMatcherInternal.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.cpp:
(WebKit::WebUserContentControllerProxy::parametersForProcess const):
(WebKit::WebUserContentControllerProxy::addStringMatcher):
(WebKit::WebUserContentControllerProxy::removeStringMatcher):
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.h:
* Source/WebKit/UIProcess/mac/WebDataListSuggestionsDropdownMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::get):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Source/WebKit/WebProcess/UserContent/SharedMemoryStringMatcher.cpp: Added.
(WebKit::SharedMemoryStringMatcher::create):
(WebKit::SharedMemoryStringMatcher::SharedMemoryStringMatcher):
* Source/WebKit/WebProcess/UserContent/SharedMemoryStringMatcher.h: Added.
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::getOrCreate):
(WebKit::WebUserContentController::removeContentWorld):
(WebKit::WebUserContentController::addStringMatcher):
(WebKit::WebUserContentController::removeStringMatcher):
(WebKit::WebUserContentController::hasStringMatchersForWorld const):
(WebKit::WebUserContentController::stringMatcher const):
* Source/WebKit/WebProcess/UserContent/WebUserContentController.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/StringMatcher.mm: Added.
(TEST(StringMatcher, MatcherData)):
(TEST(StringMatcher, MatchResults)):
(TEST(StringMatcher, IDLExposed)):
(TEST(StringMatcher, CharacterEdgeCases)):

Canonical link: <a href="https://commits.webkit.org/302111@main">https://commits.webkit.org/302111@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78481358d619e4c789a2502c104af5439179bab1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/323 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38871 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135415 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/483051f2-de79-49d9-94cc-55b806c706d9) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129917 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/201 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/97476 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7304309a-1aef-45e1-b3b8-8546d53684bc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130993 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114703 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78043 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ed028bc3-db11-45df-8883-3b183878fa7e) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/127396 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32820 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78724 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/108508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137903 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/184 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111052 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105743 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/140 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52363 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20008 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/230 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/61697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/149 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/190 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/page/WebKitStringMatcher.idl:
```diff
@@ -0,0 +1,31 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// FIXME: Don't expose this to the window.
```
Does this need to be fixed before landing? Or is it somehow not exposed in some other way?

achristensen07 replied:
Yes.  I think we made a mistake by exposing window.UserMessageHandlersNamespace and I don't want to make the same mistake here.  I think this should be fixed before merging.

weinig commented on Source/WebCore/page/WebKitStringMatcherInit.h:
```diff
@@ -0,0 +1,33 @@
+
+namespace WebCore {
+
+struct WebKitStringMatcherInit {
```
What’s the point of the empty init struct?

achristensen07 replied:
The intent to add to it in an upcoming version of this PR.  Options like stopping after the first match, searching in reverse order starting at the end of the string, etc.

weinig commented on Source/WebCore/page/WebKitStringMatcher.h:
```diff
@@ -0,0 +1,76 @@
+    struct IdentifierAndIndex;
+    Vector<IdentifierAndIndex> runDFAOnEachCharacter(const String&, const WebKitStringMatcherInit& options);
+
+    const Ref<SharedMemory> m_data;
```
Seems wrong for WebCore to be using SharedMemory directly like this, really, WebCore should not care about that kind of thing. If this data is coming from WebKit, how about making an abstract `WebKitStringMatcherData`  type that WebKit can implement using SharedMemory if it wants to.

achristensen07 replied:
CompiledContentExtension has that design, and I think it's nicer but it requires a virtual function call.  That's probably not the end of the world as long as it's not in hot inner loops.

weinig commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,277 @@
+
+namespace WebCore {
+
+RefPtr<WebKitStringMatcher> WebKitStringMatcher::create(RefPtr<SharedMemory>&& data)
```
What's the point of allowing nullptr input here?

achristensen07 replied:
This is where I do the checks to see if the data is formatted correctly.  Maybe the _WKStringMatcher constructor would be a better place to return null if it's not formatted correctly.

weinig commented on Source/WebCore/page/WebKitStringMatcher.idl:
```diff
@@ -0,0 +1,30 @@
+ */
+
+[
+    Exposed=Nowhere
```
Would `LegacyNoInterfaceObject` work here?

achristensen07 replied:
I was unaware of that.  I'll check.  Seems like it.

weinig commented on Source/WebCore/page/WebKitStringMatcherInit.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace WebCore {
+
+struct WebKitStringMatcherInit {
```
I would call this something like WebKitStringMatcherOptions, since it we usually use Init for constructors.

weinig commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,277 @@
+    return ::toJS(&globalObject, JSObjectMakeArray(context, values.size(), values.span().data(), nullptr));
+}
+
+struct Trie {
```
I'd like to see all the guts of the matcher pulled out into its own file (WebKitStringMatcherTrie.h if you will) so that WebKitStringMatcher.h/cpp can just be the binding to it.

achristensen07 replied:
I'm making it a private nested class instead.  I think that fits its role better.  It's just an intermediary when compiling, not to be used outside of WebKitStringMatcher.

weinig reviewed (commented):
I need to do a bit more thinking on this.

Right now, this seems pretty awkward, and bit too purpose built to be applicable generally.

rniwa commented on Source/WebKit/WebProcess/UserContent/WebStringMatcher.h:
```diff
@@ -0,0 +1,56 @@
+
+namespace WebKit {
+
+class WebStringMatcher final : public WebCore::WebKitStringMatcher {
```
WebStringMatcher vs. WebKitStringMatcher is a subtle difference.
Maybe we should call the WebCore version `TrieStringMatcher` or `StateMachineStringMatcher`?

achristensen07 replied:
WebStringMatcher is just the implementation that knows about shared memory.  I think we should keep the WebCore class having the same name as the IDL.  I also don't think we should have Trie in names because the input data is not necessarily from a trie.

rniwa commented on Source/WebCore/page/WebKitStringMatcher.idl:
```diff
@@ -0,0 +1,31 @@
+[
+    LegacyNoInterfaceObject,
+    SkipVTableValidation,
+] interface WebKitStringMatcher {
```
This needs to be runtime flag guarded. Or else we'd expose it in the normal world.

achristensen07 replied:
LegacyNoInterfaceObject prevents its exposure in the normal world, as verified by the StringMatcher.IDLExposed test.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/StringMatcher.mm:
```diff
@@ -0,0 +1,157 @@
+
+TEST(StringMatcher, MatcherData)
+{
+    NSMutableDictionary<NSString *, NSNumber *> *dictionary = [NSMutableDictionary dictionaryWithCapacity:26 * 26 * 26 * 4];
```
instead of hard-coding 26, can we do use `constexpr unsigned numberOfAlphabets = ('z' - 'a' + 1)`?

rniwa commented on Source/WebCore/page/WebKitStringMatcher.h:
```diff
@@ -0,0 +1,78 @@
+    struct State {
+        // a transitionsEndIndexOrMatchSentinel of 0 means transitionsBeginIndexOrMatchIdentifier is an identifier
+        // and jump to the next State without progressing through the string.
+        static constexpr uint16_t matchSentinel { 0 };
```
Hm... can we use `Markable` for this instead of manually managing a sentinel value?

achristensen07 replied:
No.  In a format with raw uint16_t's from arbitrary data we need to use a sentinel value.

rniwa commented on Source/WebCore/page/WebKitStringMatchersNamespace.h:
```diff
@@ -0,0 +1,59 @@
+    explicit WebKitStringMatchersNamespace(LocalFrame&, UserContentProvider&);
+
+    const Ref<UserContentProvider> m_userContentProvider;
+    const WeakPtr<LocalFrame> m_frame;
```
We normally back reference to a `Document`, not `Frame`.

achristensen07 replied:
This follows the pattern of UserMessageHandlersNamespace, and the UserContentProvider is accessed from the LocalFrame.

rniwa commented on Source/WebKit/WebProcess/UserContent/SharedMemoryStringMatcher.cpp:
```diff
@@ -0,0 +1,80 @@
+{
+    if (!data) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
```
We could return a trivial matcher in this case then we can change the return type of this function be `Ref` instead.

achristensen07 replied:
This should be a redundant check to verify what the UI process has already checked.  It would be difficult to test that.  I think I'm going to leave it null.

rniwa commented on Source/WebKit/WebProcess/UserContent/WebUserContentController.h:
```diff
@@ -134,6 +140,7 @@ class WebUserContentController final : public WebCore::UserContentProvider, publ
 #if ENABLE(CONTENT_EXTENSIONS)
     WebCore::ContentExtensions::ContentExtensionsBackend m_contentExtensionBackend;
 #endif
+    HashMap<ContentWorldIdentifier, HashMap<String, RefPtr<WebCore::WebKitStringMatcher>>> m_stringMatchers;
```
Doesn't this mean `StringMatcher` is effectively leaked until the document goes away? Why don't we let the client maintain a reference to it to keep it alive instead? At minimum, we should change this to WeakHashMap so that when a DOM wrapper world goes away, it'd collect the corresponding string matchers.

achristensen07 replied:
Ah, I need to add a removal in WebUserContentController::removeContentWorld

rniwa commented on Source/WebKit/Shared/WebUserContentControllerDataTypes.cpp:
```diff
@@ -0,0 +1,53 @@
+namespace WebKit {
+
+WebStringMatcherData::WebStringMatcherData(const RefPtr<WebCore::SharedMemory>& dfa, ContentWorldData&& worldData, const String& name)
+    : dfa(dfa)
```
I'd call this stringMatcherData or something since it's strictly speaking not a true DFA in the literature sense.

rniwa commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,272 @@
+        return std::nullopt;
+    auto integers = spanReinterpretCast<const uint32_t>(bytes);
+    constexpr uint32_t currentVersion { 0 };
+    constexpr size_t integersInHeader { 2 };
```
Can we call this `headerLength` or something instead? 
`integersInHeader` sounds like values which appear within the header.

rniwa commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,272 @@
+    constexpr uint32_t currentVersion { 0 };
+    constexpr size_t integersInHeader { 2 };
+    if (integers.size() < integersInHeader
+        || integers[0] != currentVersion)
```
It seems like these two expressions will fit in a single line.

rniwa commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,272 @@
+        || integers[0] != currentVersion)
+        return std::nullopt;
+    uint32_t secondInteger = integers[1];
+    size_t stateLength = static_cast<uint16_t>(secondInteger);
```
We should probably add a FIXME saying that this code needs some care if it were to be used in a big endian CPU.
Better yet, maybe spit out an error when compiling with big endian CPU. Alternatively, we can just add `bigEndianToLittleEndian2` call in `#if CPU(BIG_ENDIAN)`

achristensen07 replied:
I'll just fail completely on big-endian CPUs.  If someone wants to implement that, they can.

rniwa commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,272 @@
+            });
+            node = addResult.iterator->value;
+        }
+        node.get().lengths.append(length);
```
Maybe we can do a length check in this code so that we can avoid allocating a large data structure when either the number of node or edge overflows `uint16_t`?

achristensen07 replied:
This is actually an identifier, not a length.  Changing that.
But it already fails when serializing such a structure.  Making the constructor fail would require reworking, add complexity, and seems unnecessary.

rniwa commented on Source/WebCore/page/WebKitStringMatcher.cpp:
```diff
@@ -0,0 +1,272 @@
+    Vector<WebKitStringMatcher::State> states;
+    Vector<WebKitStringMatcher::Transition> transitions;
+
+    static constexpr uint16_t unlinkedSentinel { std::numeric_limits<uint16_t>::max() };
```
Maybe we can use something like `Expected<uint16_t, StateKind>` instead of having two free-floating sentinel values for match & unlinked?

achristensen07 replied:
A vector of optional<State> makes this better.

achristensen07 committed:
Add prototype string matcher API
https://bugs.webkit.org/show_bug.cgi?id=300829
rdar://162715982

Reviewed by Ryosuke Niwa.

Safari has some sets of strings it searches for as substrings of longer strings.
It needs to do the searches quickly with as little startup time as possible.
Arno experimented with using regular expressions instead, but that
had too much start up time because there are thousands of strings being searched
for.  I experimented with using a typical DFA with one transition per code unit,
but that had 200x as many transitions as this and didn't fit as well into a CPU's
L1 cache.  Instead, this PR introduces a trie-based finite state machine that starts
at the root with every code point in the input string.  I slightly changed the format
of the DFA from what is currently in use in Safari to create a primitive that can match
substrings of each other, such as matching "aba" and "ba" and reporting that both
were matched instead of only reporting the shortest one being matched.

This new primitive doesn't actually enable anything that can't functionally be done
with regular expressions, but for large sets of regular expressions that are used
over and over again in one process then another, this allows a reduction in time
spent compiling the regular expressions.

Even less useful than something that doesn't enable new behavior is a native API
that allows string matching on strings.  I added such an interface,
resultsForMatchingCharacters:length:matchAll:searchReverse:, but only temporarily
as a way for Safari to incrementally transition away from the injected bundle use
by having a native matcher that behaves exactly the same as _WKStringMatcher.
Once the transition is complete, this will be removed.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::shouldHaveWebKitNamespaceForWorld):
* Source/WebCore/page/UserContentController.h:
* Source/WebCore/page/UserContentProvider.h:
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::WebKitNamespace):
(WebCore::WebKitNamespace::messageHandlers):
(WebCore::WebKitNamespace::stringMatchers):
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebCore/page/WebKitStringMatcher.cpp: Added.
(WebCore::match):
(WebCore::WebKitStringMatcher::match):
(WebCore::WebKitStringMatcher::Trie::Node::traverse const):
(WebCore::WebKitStringMatcher::Trie::Trie):
(WebCore::WebKitStringMatcher::Trie::serialize const):
(WebCore::WebKitStringMatcher::dataForMatchingStrings):
(WebCore::WebKitStringMatcher::statesAndTransitionsFromVersionedData):
* Source/WebCore/page/WebKitStringMatcher.h: Copied from Source/WebKit/Shared/WebUserContentControllerDataTypes.h.
* Source/WebCore/page/WebKitStringMatcher.idl: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebCore/page/WebKitStringMatcherOptions.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebCore/page/WebKitStringMatcherOptions.idl: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebCore/page/WebKitStringMatchersNamespace.cpp: Copied from Source/WebKit/Shared/WebUserContentControllerDataTypes.h.
(WebCore::WebKitStringMatchersNamespace::WebKitStringMatchersNamespace):
(WebCore::WebKitStringMatchersNamespace::namedItem):
(WebCore::WebKitStringMatchersNamespace::supportedPropertyNames const):
(WebCore::WebKitStringMatchersNamespace::isSupportedPropertyName):
* Source/WebCore/page/WebKitStringMatchersNamespace.h: Copied from Source/WebKit/Shared/WebUserContentControllerDataTypes.h.
(WebCore::WebKitStringMatchersNamespace::create):
* Source/WebCore/page/WebKitStringMatchersNamespace.idl: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/UserContentControllerParameters.h:
* Source/WebKit/Shared/UserContentControllerParameters.serialization.in:
* Source/WebKit/Shared/WebUserContentControllerDataTypes.cpp: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
(WebKit::WebStringMatcherData::WebStringMatcherData):
(WebKit::WebStringMatcherData::sharedMemoryHandle const):
* Source/WebKit/Shared/WebUserContentControllerDataTypes.h:
* Source/WebKit/Shared/WebUserContentControllerDataTypes.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIStringMatcher.cpp: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
(API::StringMatcher::StringMatcher):
(API::StringMatcher::sharedMemory):
(API::StringMatcher::sharedBuffer):
* Source/WebKit/UIProcess/API/APIStringMatcher.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentController.mm:
(-[WKUserContentController _addStringMatcher:contentWorld:name:]):
(-[WKUserContentController _removeStringMatcherWithName:contentWorld:]):
* Source/WebKit/UIProcess/API/Cocoa/WKUserContentControllerPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKStringMatcher.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKStringMatcher.mm: Added.
(-[_WKStringMatcher initWithData:]):
(-[_WKStringMatcher initWithDataInFile:]):
(-[_WKStringMatcher dealloc]):
(+[_WKStringMatcher matcherDataForStringsAndIdentifiers:]):
(-[_WKStringMatcher resultsForMatchingCharacters:length:matchAll:searchReverse:]):
(-[_WKStringMatcher _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKStringMatcherInternal.h: Copied from Source/WebKit/Shared/UserContentControllerParameters.h.
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.cpp:
(WebKit::WebUserContentControllerProxy::parametersForProcess const):
(WebKit::WebUserContentControllerProxy::addStringMatcher):
(WebKit::WebUserContentControllerProxy::removeStringMatcher):
* Source/WebKit/UIProcess/UserContent/WebUserContentControllerProxy.h:
* Source/WebKit/UIProcess/mac/WebDataListSuggestionsDropdownMac.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::get):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Source/WebKit/WebProcess/UserContent/SharedMemoryStringMatcher.cpp: Added.
(WebKit::SharedMemoryStringMatcher::create):
(WebKit::SharedMemoryStringMatcher::SharedMemoryStringMatcher):
* Source/WebKit/WebProcess/UserContent/SharedMemoryStringMatcher.h: Added.
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::getOrCreate):
(WebKit::WebUserContentController::removeContentWorld):
(WebKit::WebUserContentController::addStringMatcher):
(WebKit::WebUserContentController::removeStringMatcher):
(WebKit::WebUserContentController::hasStringMatchersForWorld const):
(WebKit::WebUserContentController::stringMatcher const):
* Source/WebKit/WebProcess/UserContent/WebUserContentController.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/StringMatcher.mm: Added.
(TEST(StringMatcher, MatcherData)):
(TEST(StringMatcher, MatchResults)):
(TEST(StringMatcher, IDLExposed)):
(TEST(StringMatcher, CharacterEdgeCases)):

Canonical link: https://commits.webkit.org/302111@main
